spec/javascript/availability/components/ill_link.jsx.spec.js

```diff
@@ -0,0 +1,194 @@
+    '&title=book%20title&callno=123' +
+    '&rfr_id=info%3Asid%2Fcatalog.libraries.psu.edu&aulast=author%20name&date=2021');
+});
+
+test('renders a microform link', async () => {
+  const {getByRole, container} = render(
+    <IllLink holding={{locationID: 'BINDERY', callNumber: '123', libraryID: 'UP-MICRO', itemTypeID: 'MICROFORM'}} />
+  );
+
+  expect(getByRole('link')).toHaveAttribute('href', '#');
+  expect(container.getElementsByClassName('spinner-border').length).toBe(1);
+
+  await waitFor(() => expect(container.getElementsByClassName('spinner-border').length).toBe(0));
+
+  expect(getByRole('link')).toHaveTextContent('Request Scan - Penn State Users');
+
+  expect(getByRole('link')).toHaveAttribute('href',
+    'https://psu-illiad-oclc-org.ezaccess.libraries.psu.edu/illiad/upm/illiad.dll/OpenURL?' +
+    'Action=10&Form=30&isbn=1234' +
+    '&Genre=GenericRequestMicroScan&location=BINDERY' + // microform params
+    '&title=book%20title&callno=123' +
+    '&rfr_id=info%3Asid%2Fcatalog.libraries.psu.edu&aulast=author%20name&date=2021');
+});
+
+test('renders an archival thesis link', async () => {
+  const {getByRole, container} = render(
+    <IllLink holding={{locationID: 'ARKTHESES', callNumber: '123'}} />
+  );
+
+  expect(getByRole('link')).toHaveAttribute('href', '#');
+  expect(container.getElementsByClassName('spinner-border').length).toBe(1);
+
+  await waitFor(() => expect(container.getElementsByClassName('spinner-border').length).toBe(0));
+
+  expect(getByRole('link')).toHaveTextContent('Request Scan - Penn State Users');
+
+  expect(getByRole('link')).toHaveAttribute('href',
+    'https://psu-illiad-oclc-org.ezaccess.libraries.psu.edu/illiad/upm/illiad.dll/OpenURL?Action=10' +
+    '&Form=20&Genre=GenericRequestThesisDigitization' + // archival thesis params
+    '&title=book%20title&callno=123' +
+    '&rfr_id=info%3Asid%2Fcatalog.libraries.psu.edu&aulast=author%20name&date=2021');
+});
+
+test('renders a link with no author info', async () => {
+  $.get = jest.fn().mockImplementation(() => {
+    return Promise.resolve({
+      title_245ab_tsim: 'book title',
+      pub_date_illiad_ssm: 2021,
+      isbn_ssm: '1234'
+    });
+  });
+
+  const {getByRole, container} = render(
+    <IllLink holding={{locationID: 'BINDERY', callNumber: '123'}} />
+  );
+
+  expect(getByRole('link')).toHaveAttribute('href', '#');
+  expect(container.getElementsByClassName('spinner-border').length).toBe(1);
+
+  await waitFor(() => expect(container.getElementsByClassName('spinner-border').length).toBe(0));
+
+  expect(getByRole('link')).toHaveTextContent('Copy unavailable, request via Interlibrary Loan');
+
+  expect(getByRole('link')).toHaveAttribute('href',
+    'https://psu-illiad-oclc-org.ezaccess.libraries.psu.edu/illiad/upm/illiad.dll/OpenURL?' +
+    'Action=10&Form=30&isbn=1234&title=book%20title&callno=123' +
+    '&rfr_id=info%3Asid%2Fcatalog.libraries.psu.edu&date=2021');
+});
+
```

Comment:
It seems like at this point you're shifting into a new "context", to borrow a term from rspec. The `beforeEach()` you have at the top is setting up a situation where you have all fields present, then from here down, you start testing different situations where there are some fields missing. However, it still looks like your original beforeEach() is being executed.

Is there a way to segment these tests so that the ones where we want all fields are separate from the others? If I was doing this in RSpec, I'd structure it like:

``` ruby
context 'when all filelds are present' do
  before { ... }

  it ...
  it ...
end

context 'when author is missing' do
  before { "do the author missing setup" }

  it ...
end
```

But, I don't know if such a setup is possible with jest.

Reply:
That's a great point. Jest has `describe()` which is basically the same thing as `context` in RSpec. I'll make some changes to scope the tests as you've suggested.